

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>3   Usage &mdash; bdg-sequila 0.3-SNAPSHOT documentation</title>
  

  
  
  
  

  

  
  
    

  

  
  
    <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  

  

  
        <link rel="index" title="Index"
              href="../genindex.html"/>
        <link rel="search" title="Search" href="../search.html"/>
    <link rel="top" title="bdg-sequila 0.3-SNAPSHOT documentation" href="../index.html"/>
        <link rel="next" title="4   SeQuiLa API" href="../function/function.html"/>
        <link rel="prev" title="2   Overview" href="../overview/overview.html"/> 

  
  <script src="../_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav" role="document">

   
  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="../index.html" class="icon icon-home"> bdg-sequila
          

          
            
            <img src="../_static/sequila.png" class="logo" />
          
          </a>

          
            
            
              <div class="version">
                0.3-SNAPSHOT
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../quickstart/quickstart.html">1&nbsp;&nbsp;&nbsp;Quickstart</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../overview/overview.html">2&nbsp;&nbsp;&nbsp;Overview</a></li>
</ul>
<ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">3&nbsp;&nbsp;&nbsp;Usage</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#ad-hoc-analysis">3.1&nbsp;&nbsp;&nbsp;Ad-hoc analysis</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#using-predefined-scripts-in-docker-container">3.1.1&nbsp;&nbsp;&nbsp;Using predefined scripts in docker container</a></li>
<li class="toctree-l3"><a class="reference internal" href="#writing-short-analysis-in-bdg-shell">3.1.2&nbsp;&nbsp;&nbsp;Writing short analysis in bdg-shell</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#integration-with-existing-applications">3.2&nbsp;&nbsp;&nbsp;Integration with existing applications</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#integration-with-spark-application">3.2.1&nbsp;&nbsp;&nbsp;Integration with Spark-application</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#integration-with-r-using-sparkr">3.3&nbsp;&nbsp;&nbsp;Integration with R using SparkR</a></li>
<li class="toctree-l2"><a class="reference internal" href="#integration-over-jdbc-with-sequila-thrift-server">3.4&nbsp;&nbsp;&nbsp;Integration over JDBC with SeQuiLa Thrift Server</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#integration-with-r-application">3.4.1&nbsp;&nbsp;&nbsp;Integration with R-application</a></li>
<li class="toctree-l3"><a class="reference internal" href="#integration-with-generic-applications">3.4.2&nbsp;&nbsp;&nbsp;Integration with generic applications</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#running-on-yarn">3.5&nbsp;&nbsp;&nbsp;Running on YARN</a></li>
</ul>
</li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../function/function.html">4&nbsp;&nbsp;&nbsp;SeQuiLa API</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../usecases/usecases.html">5&nbsp;&nbsp;&nbsp;Usecases</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../citation/citation.html">6&nbsp;&nbsp;&nbsp;License and citation</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">bdg-sequila</a>
        
      </nav>


      
      <div class="wy-nav-content">
        <div class="rst-content">
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../index.html">Docs</a> &raquo;</li>
        
      <li>3&nbsp;&nbsp;&nbsp;Usage</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="../_sources/usage/usage.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <a href="https://github.com/ZSI-Bio/bdg-sequila"><img style="position: absolute; top: 0; right: 0; border: 0;" src="https://camo.githubusercontent.com/365986a132ccd6a44c23a9169022c0b5c890c387/68747470733a2f2f73332e616d617a6f6e6177732e636f6d2f6769746875622f726962626f6e732f666f726b6d655f72696768745f7265645f6161303030302e706e67" alt="Fork me on GitHub" data-canonical-src="https://s3.amazonaws.com/github/ribbons/forkme_right_red_aa0000.png"></a><blockquote>
<div></div></blockquote>
<div class="section" id="usage">
<h1>3&nbsp;&nbsp;&nbsp;Usage<a class="headerlink" href="#usage" title="Permalink to this headline">¶</a></h1>
<p>SeQuiLa can be used in different ways. Specifically it supports ad-hoc research, which is typically focused on quick analysis on files. Depending on your preferences you can use predefined scripts (findOverlaps and featureCounts) or write your own code snippets within spark-shell or our bdg-shell. Additionally SeQuiLa also can be used along with your existing applications written in Scala/Spark, R or any other language/platform. Please see details below.</p>
<div class="figure">
<img alt="../_images/integration.png" src="../_images/integration.png" />
</div>
<div class="section" id="ad-hoc-analysis">
<h2>3.1&nbsp;&nbsp;&nbsp;Ad-hoc analysis<a class="headerlink" href="#ad-hoc-analysis" title="Permalink to this headline">¶</a></h2>
<p>For ad-hoc analysis SeQuiLa provides two usage patterns:</p>
<div class="section" id="using-predefined-scripts-in-docker-container">
<h3>3.1.1&nbsp;&nbsp;&nbsp;Using predefined scripts in docker container<a class="headerlink" href="#using-predefined-scripts-in-docker-container" title="Permalink to this headline">¶</a></h3>
<p>In SeQuiLa’s docker image are two predefined scripts written to be executable from commandline in good-old fashion.  No need of Scala and Spark is needed.</p>
<blockquote>
<div><div class="line-block">
<div class="line"><br /></div>
</div>
</div></blockquote>
<div class="figure">
<img alt="../_images/docker.png" src="../_images/docker.png" />
</div>
<p>Sample usage of SeQuiLa wrapped in docker container’s scripts. The snippet below shows how to download sample data files into specific directory, then run the container with mounted volume. The result should appear in specified output directory.</p>
<div class="highlight-bash"><div class="highlight"><pre><span></span><span class="nb">cd</span>  /data/sequila

wget http://.../NA12878.slice.bam

wget http://.../tgp_exome_hg18.saf

docker run --rm -it -p <span class="m">4040</span>:4040 <span class="se">\</span>
   -v /data/sequila:/data <span class="se">\</span>
   -e <span class="nv">USERID</span><span class="o">=</span><span class="nv">$UID</span> -e <span class="nv">GROUPID</span><span class="o">=</span><span class="k">$(</span>id -g<span class="k">)</span> <span class="se">\</span>
   biodatageeks/bdg-sequila <span class="se">\</span>
   featureCounts -- <span class="se">\</span>
   -o /data/featureOutput -F SAF <span class="se">\</span>
   -a /data/tgp_exome_hg18.saf /data/NA12878.slice.bam
</pre></div>
</div>
<p>Parameters passed to featureCounts are divided into two parts: equivalent to parameters passed for spark-submit (master, executor-memory, driver-memory etc.: <a class="reference external" href="https://spark.apache.org/docs/latest/submitting-applications.html">https://spark.apache.org/docs/latest/submitting-applications.html</a>) and parameters passed to featureCounts itself (input files, output files, format).</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">If you are using zsh shell remember to put double-quotes (“) when specifying master local with specified number threads. <code class="docutils literal"><span class="pre">--master</span> <span class="pre">&quot;local[4]&quot;</span></code></p>
</div>
</div>
<div class="section" id="writing-short-analysis-in-bdg-shell">
<h3>3.1.2&nbsp;&nbsp;&nbsp;Writing short analysis in bdg-shell<a class="headerlink" href="#writing-short-analysis-in-bdg-shell" title="Permalink to this headline">¶</a></h3>
<p>For Scala enthusiasts - SeQuiLa provides bdg-shell which is a wrapper for spark-shell. It has extra strategy registered  and configuration already set, so it is fit for quick analysis.</p>
<blockquote>
<div><div class="line-block">
<div class="line"><br /></div>
</div>
</div></blockquote>
<div class="figure" id="id1">
<img alt="../_images/bdg-shell.png" src="../_images/bdg-shell.png" />
<p class="caption"><span class="caption-text">Sample ad-hoc analysis</span></p>
</div>
<div class="highlight-scala"><div class="highlight"><pre><span></span><span class="k">import</span> <span class="nn">htsjdk.samtools.ValidationStringency</span>
<span class="k">import</span> <span class="nn">org.apache.hadoop.io.LongWritable</span>
<span class="k">import</span> <span class="nn">org.apache.spark.SparkContext</span>
<span class="k">import</span> <span class="nn">org.apache.spark.rdd.NewHadoopRDD</span>
<span class="k">import</span> <span class="nn">org.seqdoop.hadoop_bam.</span><span class="o">{</span><span class="nc">BAMInputFormat</span><span class="o">,</span> <span class="nc">FileVirtualSplit</span><span class="o">,</span> <span class="nc">SAMRecordWritable</span><span class="o">}</span>
<span class="k">import</span> <span class="nn">org.seqdoop.hadoop_bam.util.SAMHeaderReader</span>


<span class="n">sc</span><span class="o">.</span><span class="n">hadoopConfiguration</span><span class="o">.</span><span class="n">set</span><span class="o">(</span><span class="nc">SAMHeaderReader</span><span class="o">.</span><span class="nc">VALIDATION_STRINGENCY_PROPERTY</span><span class="o">,</span> <span class="nc">ValidationStringency</span><span class="o">.</span><span class="nc">SILENT</span><span class="o">.</span><span class="n">toString</span><span class="o">)</span>
<span class="k">case</span> <span class="k">class</span> <span class="nc">PosRecord</span><span class="o">(</span><span class="n">contigName</span><span class="k">:</span><span class="kt">String</span><span class="o">,</span><span class="n">start</span><span class="k">:</span><span class="kt">Int</span><span class="o">,</span><span class="n">end</span><span class="k">:</span><span class="kt">Int</span><span class="o">)</span>

<span class="k">val</span> <span class="n">alignments</span> <span class="k">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">newAPIHadoopFile</span><span class="o">[</span><span class="kt">LongWritable</span>, <span class="kt">SAMRecordWritable</span>, <span class="kt">BAMInputFormat</span><span class="o">](</span><span class="s">&quot;/data/granges/NA12878.ga2.exome.maq.recal.bam&quot;</span><span class="o">).</span><span class="n">map</span><span class="o">(</span><span class="k">_</span><span class="o">.</span><span class="n">_2</span><span class="o">.</span><span class="n">get</span><span class="o">).</span><span class="n">map</span><span class="o">(</span><span class="n">r</span><span class="k">=&gt;</span><span class="nc">PosRecord</span><span class="o">(</span><span class="n">r</span><span class="o">.</span><span class="n">getContig</span><span class="o">,</span><span class="n">r</span><span class="o">.</span><span class="n">getStart</span><span class="o">,</span><span class="n">r</span><span class="o">.</span><span class="n">getEnd</span><span class="o">))</span>

<span class="k">val</span> <span class="n">reads</span><span class="k">=</span><span class="n">alignments</span><span class="o">.</span><span class="n">toDF</span>
<span class="n">reads</span><span class="o">.</span><span class="n">createOrReplaceTempView</span><span class="o">(</span><span class="s">&quot;reads&quot;</span><span class="o">)</span>

<span class="k">val</span> <span class="n">targets</span> <span class="k">=</span> <span class="n">spark</span><span class="o">.</span><span class="n">read</span><span class="o">.</span><span class="n">parquet</span><span class="o">(</span><span class="s">&quot;/data/granges/tgp_exome_hg18.adam&quot;</span><span class="o">)</span>
<span class="n">targets</span><span class="o">.</span><span class="n">createOrReplaceTempView</span><span class="o">(</span><span class="s">&quot;targets&quot;</span><span class="o">)</span>

<span class="k">val</span> <span class="n">query</span><span class="o">=</span><span class="s">&quot;&quot;&quot;    SELECT targets.contigName,targets.start,targets.end,count(*) FROM reads JOIN targets</span>
<span class="s">         |         ON (targets.contigName=reads.contigName</span>
<span class="s">         |         AND</span>
<span class="s">         |         CAST(reads.end AS INTEGER)&gt;=CAST(targets.start AS INTEGER)</span>
<span class="s">         |         AND</span>
<span class="s">         |         CAST(reads.start AS INTEGER)&lt;=CAST(targets.end AS INTEGER)</span>
<span class="s">         |         )</span>
<span class="s">         |         GROUP BY targets.contigName,targets.start,targets.end&quot;&quot;&quot;</span>

<span class="k">val</span> <span class="n">reads</span> <span class="k">=</span> <span class="n">spark</span><span class="o">.</span><span class="n">read</span><span class="o">.</span><span class="n">parquet</span><span class="o">(</span><span class="s">&quot;/data/granges/NA12878.ga2.exome.maq.recal.adam&quot;</span><span class="o">)</span>
<span class="n">reads</span><span class="o">.</span><span class="n">createOrReplaceTempView</span><span class="o">(</span><span class="s">&quot;reads&quot;</span><span class="o">)</span>

<span class="k">val</span> <span class="n">targets</span> <span class="k">=</span> <span class="n">spark</span><span class="o">.</span><span class="n">read</span><span class="o">.</span><span class="n">parquet</span><span class="o">(</span><span class="s">&quot;/data/granges/tgp_exome_hg18.adam&quot;</span><span class="o">)</span>
<span class="n">targets</span><span class="o">.</span><span class="n">createOrReplaceTempView</span><span class="o">(</span><span class="s">&quot;targets&quot;</span><span class="o">)</span>
<span class="n">sqlContext</span><span class="o">.</span><span class="n">sql</span><span class="o">(</span><span class="n">query</span><span class="o">)</span>
</pre></div>
</div>
</div>
</div>
<hr class="docutils" />
<div class="section" id="integration-with-existing-applications">
<h2>3.2&nbsp;&nbsp;&nbsp;Integration with existing applications<a class="headerlink" href="#integration-with-existing-applications" title="Permalink to this headline">¶</a></h2>
<p>When you already have working application supporting your analysis pipeline - you may still use SeQuiLa, substituting your existing genomic interval queries with pure SQL.</p>
<div class="section" id="integration-with-spark-application">
<h3>3.2.1&nbsp;&nbsp;&nbsp;Integration with Spark-application<a class="headerlink" href="#integration-with-spark-application" title="Permalink to this headline">¶</a></h3>
<p>When you have existing analysis pipeline in Spark ecosystem you may benefit from SeQuiLa extra strategy registered at SparkSQL level.</p>
<div class="figure align-center">
<img alt="../_images/spark-integration.png" src="../_images/spark-integration.png" />
</div>
<p>&lt;TODO&gt; opis krokow</p>
</div>
</div>
<div class="section" id="integration-with-r-using-sparkr">
<h2>3.3&nbsp;&nbsp;&nbsp;Integration with R using SparkR<a class="headerlink" href="#integration-with-r-using-sparkr" title="Permalink to this headline">¶</a></h2>
<div class="highlight-bash"><div class="highlight"><pre><span></span>docker run -e <span class="nv">USERID</span><span class="o">=</span><span class="nv">$UID</span> -e <span class="nv">GROUPID</span><span class="o">=</span><span class="k">$(</span>id -g<span class="k">)</span> -it -v /Users/biodatageek/data:/data <span class="se">\</span>
-p <span class="m">4040</span>:4040 biodatageeks/bdg-sequila bdg-sequilaR
</pre></div>
</div>
<div class="highlight-R"><div class="highlight"><pre><span></span><span class="c1">#register SeQuilaR extensions</span>
sparkR.callJStatic<span class="p">(</span><span class="s">&quot;org.biodatageeks.R.SequilaR&quot;</span><span class="p">,</span><span class="s">&quot;init&quot;</span><span class="p">,</span>spark<span class="p">)</span>
<span class="c1">#create db</span>
sql<span class="p">(</span><span class="s">&quot;CREATE DATABASE sequila&quot;</span><span class="p">)</span>
sql<span class="p">(</span><span class="s">&quot;USE sequila&quot;</span><span class="p">)</span>
<span class="c1">#create a BAM data source with reads</span>
sql<span class="p">(</span><span class="s">&#39;CREATE TABLE reads USING org.biodatageeks.datasources.BAM.BAMDataSource OPTIONS(path &quot;/data/c1_10M.bam&quot;)&#39;</span><span class="p">)</span>
<span class="c1">#parse GTF with target regions</span>
sql<span class="p">(</span><span class="s">&#39;CREATE TABLE targets_temp(Chr string, TypeDB string, Feature string, Start integer,</span>
<span class="s">End integer, t1 string, Strand string, t2 string, Gene_id_temp string ,Gene_id string)</span>
<span class="s"> USING csv</span>
<span class="s"> OPTIONS (path &quot;/data/Homo_sapiens.gtf&quot;, header &quot;false&quot;, inferSchema &quot;false&quot;, delimiter &quot;\t&quot;)&#39;</span><span class="p">)</span>

<span class="c1">#a query to compute counts per targer</span>
query <span class="o">&lt;-</span> <span class="s">&quot;SELECT Gene_id,Chr ,targets.Start ,targets.End ,Strand ,CAST(targets.End AS INTEGER)-</span>
<span class="s">CAST(targets.Start AS INTEGER) + 1 AS Length, count(*) AS Counts FROM reads JOIN targets_temp as targets</span>
<span class="s">ON (Chr=reads.contigName AND reads.end &gt;= CAST(targets.Start AS INTEGER)</span>
<span class="s">AND reads.start &lt;= CAST(targets.End AS INTEGER)) GROUP BY Gene_id, Chr, targets.Start, targets.End, Strand&quot;</span>

<span class="c1">#check physical execution plan to verify if IntervalTreeJoinOptimChromosome strategy is used</span>
explain<span class="p">(</span>sql<span class="p">(</span>query<span class="p">))</span>
<span class="c1">#get sample output</span>
<span class="kp">head</span><span class="p">(</span>sql<span class="p">(</span>query<span class="p">))</span>

      Gene_id Chr     Start       End Strand Length Counts
<span class="m">1</span>     g1   <span class="m">6</span>  <span class="m">73263359</span>  <span class="m">73301401</span>      <span class="o">+</span>  <span class="m">38043</span>    <span class="m">157</span>
<span class="m">2</span>     g2   <span class="m">7</span>   <span class="m">6469654</span>   <span class="m">6484149</span>      <span class="o">-</span>  <span class="m">14496</span>     <span class="m">95</span>
<span class="m">3</span>     g3  <span class="m">10</span> <span class="m">123171535</span> <span class="m">123171875</span>      <span class="o">-</span>    <span class="m">341</span>    <span class="m">309</span>
<span class="m">4</span>     g4  <span class="m">15</span>  <span class="m">82540426</span>  <span class="m">82540456</span>      <span class="o">-</span>     <span class="m">31</span>    <span class="m">272</span>
<span class="m">5</span>     g5  <span class="m">20</span>  <span class="m">58891302</span>  <span class="m">58911192</span>      <span class="o">+</span>  <span class="m">19891</span>   <span class="m">6728</span>
<span class="m">6</span>     g6   <span class="m">7</span>  <span class="m">42935021</span>  <span class="m">42935136</span>      <span class="o">+</span>    <span class="m">116</span>     <span class="m">64</span>
</pre></div>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">For more detailed instruction on how to work with SparkR API please consult <a class="reference external" href="https://spark.apache.org/docs/2.3.0/sparkr.html">SparkR</a> documentation.</p>
</div>
</div>
<div class="section" id="integration-over-jdbc-with-sequila-thrift-server">
<h2>3.4&nbsp;&nbsp;&nbsp;Integration over JDBC with SeQuiLa Thrift Server<a class="headerlink" href="#integration-over-jdbc-with-sequila-thrift-server" title="Permalink to this headline">¶</a></h2>
<p>In order to start SeQuiLa Spark Thrift server you can use the following prodecure:</p>
<div class="highlight-bash"><div class="highlight"><pre><span></span>docker run --rm -p <span class="m">4040</span>:4040 -p <span class="m">12000</span>:12000 -e <span class="nv">USERID</span><span class="o">=</span><span class="nv">$UID</span> -e <span class="nv">GROUPID</span><span class="o">=</span><span class="k">$(</span>id -g<span class="k">)</span> <span class="se">\</span>
-it biodatageeks/bdg-sequila bash
bdg-start-thriftserver --hiveconf hive.server2.thrift.port<span class="o">=</span><span class="m">12000</span>
</pre></div>
</div>
<p>Once done simply stop it as follows:</p>
<div class="highlight-bash"><div class="highlight"><pre><span></span>bdg-stop-thriftserver
</pre></div>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">For detailed instructions on how to run Spark Thrift Server please check this <a class="reference external" href="https://developer.ibm.com/hadoop/2016/08/22/how-to-run-queries-on-spark-sql-using-jdbc-via-thrift-server/">page</a>.
Please note that all options including resource management can be set in exactly the same way as in Spark Thrift Server.</p>
</div>
<div class="section" id="integration-with-r-application">
<h3>3.4.1&nbsp;&nbsp;&nbsp;Integration with R-application<a class="headerlink" href="#integration-with-r-application" title="Permalink to this headline">¶</a></h3>
<ol class="arabic simple">
<li>Install rJava and RJDBC packages:</li>
</ol>
<div class="highlight-R"><div class="highlight"><pre><span></span>install.packages<span class="p">(</span><span class="s">&quot;RJDBC&quot;</span><span class="p">,</span>dep<span class="o">=</span><span class="kc">TRUE</span><span class="p">)</span>
install.packages<span class="p">(</span><span class="s">&quot;rJava&quot;</span><span class="p">)</span>
<span class="kn">library</span><span class="p">(</span>RJDBC<span class="p">)</span>
</pre></div>
</div>
<ol class="arabic simple" start="2">
<li>Download Spark JDBC driver - for the convenience we have already prepare a self-contained jar file for you:</li>
</ol>
<div class="highlight-R"><div class="highlight"><pre><span></span>download.file<span class="p">(</span><span class="s">&quot;http://zsibio.ii.pw.edu.pl/nexus/repository/maven-releases/org/biodatageeeks/spark/jdbc/spark-jdbc_2.11/0.12/spark-jdbc_2.11-0.12-assembly.jar&quot;</span><span class="p">,</span>destfile <span class="o">=</span> <span class="s">&quot;spark-jdbc-assembly-0.12.jar&quot;</span><span class="p">)</span>
</pre></div>
</div>
<ol class="arabic simple" start="3">
<li>Establish a connection to the Spark Thrift Server you have started in the previous section:</li>
</ol>
<div class="highlight-R"><div class="highlight"><pre><span></span>drv <span class="o">&lt;-</span> JDBC<span class="p">(</span><span class="s">&quot;org.apache.hive.jdbc.HiveDriver&quot;</span><span class="p">,</span>classPath <span class="o">=</span> <span class="s">&quot;./spark-jdbc-assembly-0.12.jar&quot;</span><span class="p">,</span>identifier.quote<span class="o">=</span><span class="s">&quot;`&quot;</span><span class="p">)</span>
conn <span class="o">&lt;-</span> dbConnect<span class="p">(</span>drv<span class="p">,</span> <span class="s">&quot;jdbc:hive2://localhost:12000&quot;</span><span class="p">,</span> <span class="s">&quot;user&quot;</span><span class="p">,</span> <span class="s">&quot;passord&quot;</span><span class="p">)</span>


ds <span class="o">&lt;-</span>dbGetQuery<span class="p">(</span>conn<span class="p">,</span> <span class="s">&quot;SELECT targets.GeneId AS GeneId,</span>
<span class="s">                     targets.Chr AS Chr,</span>
<span class="s">                targets.Start AS Start,</span>
<span class="s">                targets.End AS End,</span>
<span class="s">                targets.Strand AS Strand,</span>
<span class="s">                CAST(targets.End AS INTEGER)-CAST(targets.Start AS INTEGER) + 1 AS Length,</span>
<span class="s">                count(*) AS Counts</span>
<span class="s">                FROM granges.NA12878_marek reads JOIN granges.targets targets</span>
<span class="s">                ON (</span>
<span class="s">                targets.Chr=reads.contigName</span>
<span class="s">                AND</span>
<span class="s">                reads.end &gt;= CAST(targets.Start AS INTEGER)</span>
<span class="s">                AND</span>
<span class="s">                reads.start &lt;= CAST(targets.End AS INTEGER)</span>
<span class="s">                )</span>
<span class="s">                GROUP BY targets.GeneId,targets.Chr,targets.Start,targets.End,targets.Strand&quot;</span><span class="p">)</span>

<span class="kp">nrow</span><span class="p">(</span>ds<span class="p">)</span>
<span class="kp">head</span><span class="p">(</span>ds<span class="p">)</span>
dbDisconnect<span class="p">(</span>conn<span class="p">)</span>
</pre></div>
</div>
<p>Once done you should be able to see a similar result on your screen:</p>
<img alt="../_images/rstudio.png" src="../_images/rstudio.png" />
</div>
<div class="section" id="integration-with-generic-applications">
<h3>3.4.2&nbsp;&nbsp;&nbsp;Integration with generic applications<a class="headerlink" href="#integration-with-generic-applications" title="Permalink to this headline">¶</a></h3>
<p>When integrating SeQuiLa with generic, non-Spark, non-R application you need additional component (namely Spark Thrift Server) with injected SeQuiLa strategy which is exposing JDBC/ODBC interface. Afterwards you can connect to Thrift Server through JDBC interface, load data and query it in SQL language.</p>
<div class="figure align-center">
<img alt="../_images/thrift-server.png" src="../_images/thrift-server.png" />
</div>
<p>We will show how JDBC integration works with one of the SQL client, for example: <a class="reference external" href="http://www.squirrelsql.org/">http://www.squirrelsql.org/</a></p>
<p>At your favourite SQL client setup connection to Spark Thrift Server.</p>
<p>You will need Spark JDBC driver. We have prepared assembly jar for this purpose: <a class="reference external" href="http://zsibio.ii.pw.edu.pl/nexus/repository/maven-releases/org/biodatageeeks/spark/jdbc/spark-jdbc_2.11/0.12/spark-jdbc_2.11-0.12-assembly.jar">http://zsibio.ii.pw.edu.pl/nexus/repository/maven-releases/org/biodatageeeks/spark/jdbc/spark-jdbc_2.11/0.12/spark-jdbc_2.11-0.12-assembly.jar</a></p>
<p>For example in Squirrel SQL configure new driver:</p>
<div class="figure align-center">
<img alt="../_images/jdbc.png" src="../_images/jdbc.png" />
</div>
<p>Create new Alias:</p>
<div class="figure align-center">
<a class="reference internal image-reference" href="../_images/alias.png"><img alt="../_images/alias.png" src="../_images/alias.png" style="width: 342.0px; height: 347.0px;" /></a>
</div>
<p>Afterwards you can play with SQL.</p>
<div class="highlight-sql"><div class="highlight"><pre><span></span><span class="c1">---reads</span>
<span class="k">CREATE</span> <span class="k">TABLE</span> <span class="n">granges</span><span class="p">.</span><span class="n">NA12878_marek</span>
<span class="k">USING</span> <span class="n">org</span><span class="p">.</span><span class="n">biodatageeks</span><span class="p">.</span><span class="n">datasources</span><span class="p">.</span><span class="n">BAM</span><span class="p">.</span><span class="n">BAMDataSource</span>
<span class="k">OPTIONS</span><span class="p">(</span><span class="n">path</span> <span class="ss">&quot;/data/granges/NA12878.ga2.exome.maq.recal.bam&quot;</span><span class="p">);</span>

<span class="c1">--targets</span>
<span class="k">CREATE</span> <span class="k">TABLE</span> <span class="n">granges</span><span class="p">.</span><span class="n">targets</span>
<span class="k">USING</span> <span class="n">csv</span>
<span class="k">OPTIONS</span> <span class="p">(</span><span class="n">path</span> <span class="ss">&quot;/data/granges/tgp_exome_hg18.saf&quot;</span><span class="p">,</span> <span class="n">header</span> <span class="ss">&quot;true&quot;</span><span class="p">,</span> <span class="n">inferSchema</span> <span class="ss">&quot;false&quot;</span><span class="p">,</span> <span class="k">delimiter</span> <span class="ss">&quot;\t&quot;</span><span class="p">);</span>

<span class="k">SELECT</span> <span class="k">count</span><span class="p">(</span><span class="o">*</span><span class="p">)</span> <span class="k">from</span> <span class="n">granges</span><span class="p">.</span><span class="n">NA12878_marek</span><span class="p">;</span>
<span class="k">SELECT</span> <span class="k">count</span><span class="p">(</span><span class="o">*</span><span class="p">)</span> <span class="k">from</span> <span class="n">granges</span><span class="p">.</span><span class="n">targets</span> <span class="k">limit</span> <span class="mi">1</span><span class="p">;</span>


<span class="k">SELECT</span> <span class="n">targets</span><span class="p">.</span><span class="n">GeneId</span> <span class="k">AS</span> <span class="n">GeneId</span><span class="p">,</span>
                     <span class="n">targets</span><span class="p">.</span><span class="n">Chr</span> <span class="k">AS</span> <span class="n">Chr</span><span class="p">,</span>
                     <span class="n">targets</span><span class="p">.</span><span class="k">Start</span> <span class="k">AS</span> <span class="k">Start</span><span class="p">,</span>
                     <span class="n">targets</span><span class="p">.</span><span class="k">End</span> <span class="k">AS</span> <span class="k">End</span><span class="p">,</span>
                     <span class="n">targets</span><span class="p">.</span><span class="n">Strand</span> <span class="k">AS</span> <span class="n">Strand</span><span class="p">,</span>
                     <span class="k">CAST</span><span class="p">(</span><span class="n">targets</span><span class="p">.</span><span class="k">End</span> <span class="k">AS</span> <span class="nb">INTEGER</span><span class="p">)</span><span class="o">-</span><span class="k">CAST</span><span class="p">(</span><span class="n">targets</span><span class="p">.</span><span class="k">Start</span> <span class="k">AS</span> <span class="nb">INTEGER</span><span class="p">)</span> <span class="o">+</span> <span class="mi">1</span> <span class="k">AS</span> <span class="k">Length</span><span class="p">,</span>
                     <span class="k">count</span><span class="p">(</span><span class="o">*</span><span class="p">)</span> <span class="k">AS</span> <span class="n">Counts</span>
            <span class="k">FROM</span> <span class="n">granges</span><span class="p">.</span><span class="n">NA12878_marek</span> <span class="k">reads</span> <span class="k">JOIN</span> <span class="n">granges</span><span class="p">.</span><span class="n">targets</span> <span class="n">targets</span>
<span class="k">ON</span> <span class="p">(</span>
  <span class="n">targets</span><span class="p">.</span><span class="n">Chr</span><span class="o">=</span><span class="k">reads</span><span class="p">.</span><span class="n">contigName</span>
  <span class="k">AND</span>
  <span class="k">reads</span><span class="p">.</span><span class="k">end</span> <span class="o">&gt;=</span> <span class="k">CAST</span><span class="p">(</span><span class="n">targets</span><span class="p">.</span><span class="k">Start</span> <span class="k">AS</span> <span class="nb">INTEGER</span><span class="p">)</span>
  <span class="k">AND</span>
  <span class="k">reads</span><span class="p">.</span><span class="k">start</span> <span class="o">&lt;=</span> <span class="k">CAST</span><span class="p">(</span><span class="n">targets</span><span class="p">.</span><span class="k">End</span> <span class="k">AS</span> <span class="nb">INTEGER</span><span class="p">)</span>
<span class="p">)</span>
<span class="k">GROUP</span> <span class="k">BY</span> <span class="n">targets</span><span class="p">.</span><span class="n">GeneId</span><span class="p">,</span><span class="n">targets</span><span class="p">.</span><span class="n">Chr</span><span class="p">,</span><span class="n">targets</span><span class="p">.</span><span class="k">Start</span><span class="p">,</span><span class="n">targets</span><span class="p">.</span><span class="k">End</span><span class="p">,</span><span class="n">targets</span><span class="p">.</span><span class="n">Strand</span><span class="p">;</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="running-on-yarn">
<h2>3.5&nbsp;&nbsp;&nbsp;Running on YARN<a class="headerlink" href="#running-on-yarn" title="Permalink to this headline">¶</a></h2>
</div>
</div>


           </div>
           <div class="articleComments">
            
           </div>
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="../function/function.html" class="btn btn-neutral float-right" title="4   SeQuiLa API" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="../overview/overview.html" class="btn btn-neutral" title="2   Overview" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2018, biodatageeks.org.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'../',
            VERSION:'0.3-SNAPSHOT',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true,
            SOURCELINK_SUFFIX: '.txt'
        };
    </script>
      <script type="text/javascript" src="../_static/jquery.js"></script>
      <script type="text/javascript" src="../_static/underscore.js"></script>
      <script type="text/javascript" src="../_static/doctools.js"></script>

  

  
  
    <script type="text/javascript" src="../_static/js/theme.js"></script>
  

  
  
  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.StickyNav.enable();
      });
  </script>
   

</body>
</html>